spec/unit/pops/lookup/context_spec.rb

```diff
@@ -0,0 +1,120 @@
+#! /usr/bin/env ruby
+require 'spec_helper'
+require 'puppet_spec/compiler'
+
+module Puppet::Pops
+module Lookup
+describe 'Puppet::Pops::Lookup::Context' do
+
+  context 'an instance' do
+    include PuppetSpec::Compiler
+    it 'can be created' do
+      code = "notice(type(Puppet::LookupContext.new('e', 'm')))"
+      expect(eval_and_collect_notices(code)[0]).to match(/Object\[\{name => 'Puppet::LookupContext'/)
+    end
+
+    it 'returns its environment_name' do
+      code = "notice(Puppet::LookupContext.new('e', 'm').environment_name)"
+      expect(eval_and_collect_notices(code)[0]).to eql('e')
+    end
+
+    it 'returns its module_name' do
+      code = "notice(Puppet::LookupContext.new('e', 'm').module_name)"
+      expect(eval_and_collect_notices(code)[0]).to eql('m')
+    end
+
+    it 'can use an undef module_name' do
+      code = "notice(type(Puppet::LookupContext.new('e', undef).module_name))"
+      expect(eval_and_collect_notices(code)[0]).to eql('Undef')
+    end
+
+    it 'can store and retrieve a value using the cache' do
+      code = <<-PUPPET.unindent
+        $ctx = Puppet::LookupContext.new('e', 'm')
+        $ctx.cache('ze_key', 'ze_value')
+        notice($ctx.cached_value('ze_key'))
+      PUPPET
+      expect(eval_and_collect_notices(code)[0]).to eql('ze_value')
+    end
+
+    it 'can store and retrieve a hash using the cache' do
+      code = <<-PUPPET.unindent
+        $ctx = Puppet::LookupContext.new('e', 'm')
+        $ctx.cache_all({ 'v1' => 'first', 'v2' => 'second' })
+        $ctx.cached_entries |$key, $value| { notice($key); notice($value) }
+      PUPPET
+      expect(eval_and_collect_notices(code)).to eql(%w(v1 first v2 second))
+    end
+
+    it 'can use the cache to merge hashes' do
+      code = <<-PUPPET.unindent
+        $ctx = Puppet::LookupContext.new('e', 'm')
+        $ctx.cache_all({ 'v1' => 'first', 'v2' => 'second' })
+        $ctx.cache_all({ 'v3' => 'third', 'v4' => 'fourth' })
+        $ctx.cached_entries |$key, $value| { notice($key); notice($value) }
+      PUPPET
+      expect(eval_and_collect_notices(code)).to eql(%w(v1 first v2 second v3 third v4 fourth))
+    end
+
+    it 'can use the cache to merge hashes and individual entries' do
+      code = <<-PUPPET.unindent
+        $ctx = Puppet::LookupContext.new('e', 'm')
+        $ctx.cache_all({ 'v1' => 'first', 'v2' => 'second' })
+        $ctx.cache('v3', 'third')
+        $ctx.cached_entries |$key, $value| { notice($key); notice($value) }
+      PUPPET
+      expect(eval_and_collect_notices(code)).to eql(%w(v1 first v2 second v3 third))
+    end
+
```

Comment:
test missing: can replace a value for a cache key `cache('x', 10); notice cached_value('x'); cache('x', 20); notice cached_value('x');`